File: DSA/src/Questions/String/MultiplyStrings.java
package Questions.String;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/*
https://leetcode.com/problems/multiply-strings/description/
 Given two non-negative integers num1 and num2 represented as strings, return the product of num1 and num2, also represented as a string.

Note: You must not use any built-in BigInteger library or convert the inputs to integer directly.

 

Example 1:

Input: num1 = "2", num2 = "3"
Output: "6"
Example 2:

Input: num1 = "123", num2 = "456"
Output: "56088"
 

Constraints:

1 <= num1.length, num2.length <= 200
num1 and num2 consist of digits only.
Both num1 and num2 do not contain any leading zero, except the number 0 itself.
Time Complexity: O(M*N)
Space Complexity: O(M + N + 1)
 */
public class MultiplyStrings {
    public static String multiply(String num1, String num2) {
        if ("0".equals(num1) || "0".equals(num2)) {
            return "0";
        }
        int m = num1.length(), n = num2.length();
        List<Integer> res = new ArrayList<>(Collections.nCopies(
                m + n, 0)); // Initialize the result list with zeros

        // Reverse iterate through num1 and num2
        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                int mul = (num1.charAt(i) - '0') *
                        (num2.charAt(j) - '0'); // Multiply digits
                int p1 = i + j, p2 = i + j + 1; // Position in the result list
                int sum = mul + res.get(p2); // Add to the current position

                res.set(p1, res.get(p1) + sum / 10); // Carry
                res.set(p2, sum % 10); // Remainder
            }
        }

        // Build the result string
        StringBuilder sb = new StringBuilder();
        for (int p : res) {
            if (!(sb.length() == 0 && p == 0)) { // Skip leading zeros
                sb.append(p);
            }
        }

        return sb.length() == 0
                ? "0"
                : sb.toString(); // Return result or "0" if empty
    }

    public static void main(String[] args) {
        // Example usage
        System.out.println(multiply("123", "456")); // Expected output: "56088"
    }
}

File: DSA/Python/Questions/String/MultiplyStrings.py
"""
# Question: Multiply Strings
# Link: https://leetcode.com/problems/multiply-strings/

# Time Complexity: O(m * n)
# Space Complexity: O(m + n)

# Algorithm:
# 1. Process digits from right to left
# 2. Store intermediate results
# 3. Handle carry values
# 4. Build final result string

# Key Components:
# - multiply(): Main implementation
# - Digit-by-digit multiplication
# - Result array management
"""


class MultiplyStrings:
    def multiply(self, num1: str, num2: str) -> str:
        if num1 == "0" or num2 == "0":
            return "0"

        m, n = len(num1), len(num2)
        result = [0] * (m + n)

        # Multiply each digit
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                mul = (ord(num1[i]) - ord("0")) * (ord(num2[j]) - ord("0"))
                p1, p2 = i + j, i + j + 1
                total = mul + result[p2]

                result[p1] += total // 10
                result[p2] = total % 10

        # Build result string
        final = []
        i = 0
        while i < len(result) and result[i] == 0:
            i += 1

        while i < len(result):
            final.append(str(result[i]))
            i += 1

        return "".join(final) if final else "0"


def main():
    solution = MultiplyStrings()
    print(solution.multiply("123", "456"))  # Expected: "56088"


if __name__ == "__main__":
    main()
